docs/proposals/deployment.md

```diff
@@ -0,0 +1,172 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/deployment.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Deployment
+
+## Abstract
+
+A proposal for implementing a new resource - Deployment - which will enable
+declarative config updates for Pods and ReplicationControllers.
+
+Users will be able to create a Deployment, which will spin of
+a ReplicationController to bring up the desired pods.
+Users can also target the Deployment at existing ReplicationControllers, in
+which case the new RC will replace the existing ones. The exact mechanics of
+replacement depends on the DeploymentStrategy chosen by the user.
+DeploymentStrategies are explained in detail in a later section.
+
+## Implementation
+
+### API Object
+
+The `Deployment` API object will have the following structure:
+
+```go
+type Deployment struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior of the Deployment.
+  Spec DeploymentSpec
+
+  // Most recently observed status of the Deployment.
+  Status DeploymentStatus
+}
+
+type DeploymentSpec struct {
+  // Number of desired pods. This is a pointer to distinguish between explicit
+  // zero and not specified. Defaults to 1.
+  Replicas *int
+
+  // Label selector for pods. Existing ReplicationControllers which have all these
+  // label selectors will be scaled down.
+  Selector map[string][string]
+
+  // Describes the pods that will be created.
+  Template *PodTemplateSpec
+
+  // The deployment strategy to use to replace existing pods with new ones.
+  Strategy DeploymentStrategy
+}
+
+type DeploymentStrategy struct {
+  // Type of deployment. Can be "Recreate" or "RollingUpdate".
+  Type DeploymentType
```

Comment:
On Aug 11, 2015, at 10:29 PM, Brian Grant <notifications@github.com> wrote:

In docs/proposals/deployment.md
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12236#discussion_r36821874>
:

> +  Replicas *int
> +
> +  // Label selector for pods. Existing ReplicationControllers which have all these
> +  // label selectors will be scaled down.
> +  Selector map[string][string]
> +
> +  // Describes the pods that will be created.
> +  Template *PodTemplateSpec
> +
> +  // The deployment strategy to use to replace existing pods with new ones.
> +  Strategy DeploymentStrategy
> +}
> +
> +type DeploymentStrategy struct {
> +  // Type of deployment. Can be "Recreate" or "RollingUpdate".
> +  Type DeploymentType

Union/OneOf is something we should add to the API conventions.

We have feedback that undiscriminated unions are problematic: #6979
<https://github.com/GoogleCloudPlatform/kubernetes/issues/6979>.


Our UI team has complained because of that as well, also, the Type field
being CamelCase isn't much better because they can't use the value as a
field index for the type, but instead have to use a switch or change the
case.  Ie "strategy[strategy.type]" in JS doesn't return
"strategy.rollingUpdate" because of case.


We could just put a TODO here to follow the convention, whatever we decide
it should be.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12236/files#r36821874>
.